### ISSUE TITLE
include passed bindings when "Error: Undefined binding(s)" is thrown

### ISSUE THREAD

**Issue by acarl005** [tgriesser/knex#2295]
We've had issues in our code which causes this error:
```
Error: Undefined binding(s) detected when compiling RAW query
<SQL query with some ? bindings>
```

We know the problem is one of the values of the bindings is `undefined`. However, we don't know which one. It would improve the debugging experience if the error message that was thrown includes the values of the bindings. Something like this perhaps:

```
Error: Undefined binding(s) detected when compiling RAW query
<SQL query with some ? bindings>
bindings: [ 'foo', 'bar', undefined, 1 ]
```

Then we would know which one of our bindings was the culprit.

**Comment by richraid21** [comment]
I'm not sure this is a good idea. [Exposing possible sensitive information through errors/logs](https://cwe.mitre.org/data/definitions/209.html) should be avoided. 

Of course, open for discussion!

**Comment by maxstoyanov** [comment]
This behavior could be enabled with the debug switch. Since debug mode is usually not engaged in production environments important information would not be leaked into logs.

**Comment by richraid21** [comment]
`KNEX_MODE= dev | prod` ?

**Comment by acarl005** [comment]
Agreed that persisting the bindings could expose sensitive data. It certainly should not be the default behavior. Perhaps something like `DEBUG=knex:bindings` to enable it?

**Comment by wubzz** [comment]
Definitely only under a debug flag.

Currently `DEBUG=knex:bindings` logs the bindings: https://github.com/tgriesser/knex/blob/master/src/query/compiler.js#L75

I think that's a bit stale and should also include the sql.

There are two additional sources of information that can be used:

**Fully compiled SQL**: Use `this.client._formatQuery(val.sql, defaults.bindings)` *(+Add support [here](https://github.com/tgriesser/knex/blob/master/src/query/string.js#L31) to preserve the undefined value)*

**Column name**: Can be picked up [here](https://github.com/tgriesser/knex/blob/master/src/query/compiler.js#L276) via `stmt.column`

Maybe I'm overthinking it. The bindings alone is probably enough.

**Comment by wubzz** [comment]
Instead of spewing this out to the console / forcing the use of debug flags, what about one of the following?

- `onUndefinedBindings` in knex config - a function that receives the builder, the sql, and the bindings.
- An event `onUndefinedBindings` - an event that receives the builder, the sql, the bindings.

This way knex doesn't spew out sensitive information. The responsibility falls completely on the user how they wish to handle this occurance. Knex still also has to throw the actual `Error` though.

**Comment by boylove142** [comment]
Any update on this? Currently, for me, the error message is really not helpful in detecting errors. 

I think we could enable/disable this feature using some config e.g. turning `secureError: false` (default secureError is true)

**Comment by elhigu** [comment]
@boylove if you are getting that error and you dont know where you are passing undefined parameters, then you should definitely not turn those errors off 😮 also I wouldnt want to even allow to globally configure anything like that. We haven’t even defined what different methods should do when undefined is passed to api. Some method for building query that allows to pass undefined to where clauses like .skipUndefinedInWhere could make sense though.

**Comment by jpike88** [comment]
This is a major pain in the butt, I have many queries that take in 3-5 bindings, and instead of playing whack a mole I'd like the option to turn on the binding values.

**Comment by kibertoad** [comment]
@jpike88 PR is welcome!

**Comment by KristjanTammekivi** [comment]
How about if bindings is array, it shows what indices are undefined, and if bindings is a dictionary, it shows what keys are undefined? This would solve the no-sensitive-data issue and it would be appropriate for production?

**Comment by elhigu** [comment]
@jpike88 You can write knex.on('query', ...) hook which prints your query SQL and bindings. I don't think there is any reason to add new feature for this. It is trivial also to setup hook in a way that it prints error only on undefined bindings.

@KristjanTammekivi That is a really good idea, I would merge PR for that 👍

**Comment by mercmobily** [comment]
Hello! Thanks for raising this issue back in 2017. I'd like to provide an update on the current state of this feature request.


**Current Status: Partially Implemented**

This issue has been partially addressed but not fully resolved. Let me explain what's been done and what remains.


**What Has Been Implemented**

PR #3425 added support for showing which columns contain undefined values **when they appear in WHERE clauses**. The error message now looks like this:

```
Undefined binding(s) detected when compiling SELECT. Undefined column(s): [username, status] query: select * from `users` where `username` = ? and `status` = ?
```

This was released in version 0.19.4 (you can see it in lib/query/querycompiler.js:110-117).


For RAW queries with array or object bindings, the indices/keys of undefined bindings are also shown (lib/raw.js:106-112):

```
Undefined binding(s) detected for keys [0, 2] when compiling RAW query: insert into `test` values (?, ?, ?)
```

**What's Still Missing**

The original request asked for the **actual binding values** to be shown in the error message, not just which indices/keys are undefined. This was discussed extensively in the thread, with several concerns raised:

1. **Security concerns**: Exposing binding values in error messages could leak sensitive information (passwords, API keys, personal data) into error logs. This is a legitimate security issue (CWE-209).

2. **Proposed solutions that were discussed**:
   - Using `DEBUG=knex:bindings` environment variable (already exists)
   - A configuration option like `secureError: false`
   - An `onUndefinedBindings` callback or event
   - A `KNEX_MODE=dev|prod` setting

3. **What maintainer @elhigu suggested**: Use the `knex.on('query', ...)` hook to print queries and bindings, which can be configured to only print on undefined binding errors.


**Current Workarounds**

There are two good workarounds available today:

1. **Enable debug mode** (shows all bindings for all queries):
```javascript
DEBUG=knex:bindings node your-app.js
```

2. **Enable async stack traces** (recommended for development):
```javascript
const knex = require('knex')({
  client: 'mysql',
  asyncStackTraces: true,  // Shows better stack traces
  connection: { /* ... */ }
});
```

The `asyncStackTraces` option (available since v0.21.0) significantly improves error stack traces, making it much easier to locate the source of undefined bindings in your code. This addresses the "useless stack trace" problem mentioned by @jpike88.

3. **Query event listener** (custom solution):
```javascript
let lastQuery = null;
knex.on('query', (query) => {
  lastQuery = query;
});

// In your error handler:
if (error.message.includes('Undefined binding')) {
  console.log('Query:', lastQuery.sql);
  console.log('Bindings:', lastQuery.bindings);
}
```


**Why This Remains Incomplete**

The maintainers have been cautious about automatically including binding values in error messages due to security implications. Even though the error is helpful during development, in production environments these errors would be logged, potentially exposing sensitive data.

The community hasn't reached consensus on the best approach - whether it should be:
- An opt-in configuration option
- A callback/event system
- Kept as-is with better documentation of workarounds


**Recommendation Going Forward**

I believe this issue deserves to remain open, but perhaps be reframed as a feature request for a **secure, opt-in mechanism** to include binding values in undefined binding errors. Something like:

```javascript
const knex = require('knex')({
  client: 'mysql',
  connection: { /* ... */ },
  debug: false,
  asyncStackTraces: process.env.NODE_ENV === 'development',
  verboseUndefinedBindings: process.env.NODE_ENV === 'development',  // proposed
});
```

This would make the feature discoverable and controllable, whilst avoiding security issues in production.


**Related Issues**

- #3723 - Strange "Undefined binding(s) detected" error message (edge case with raw queries wrapping inserts)
- #4531 - Thrown undefined bindings promise does not carry the full stack in the error (stack trace issue, partially resolved by asyncStackTraces)
- #1557 - Improve bindings debug information (closed, led to the current implementation)
- #3548 - Inserting a JSON value with an undefined property via knex.raw causes an exception
- #2427 - Console.log() should get SQL string by default (related to debugging experience)
